content/news/2021-04-04-bevy-0.5/index.md

```diff
@@ -0,0 +1,1471 @@
+        .with_system(bar.system())
+    )
+```
+
+### Improved Run Criteria
+
+Run Criteria are now decoupled from systems and will be re-used when possible. For example, the FixedTimestep criteria in the example above will only be run once per stage run. The executor will re-use the criteria's result for both the `foo` and `bar` system.
+
+Run Criteria can now also be labeled and referenced by other systems:
+
+
+```rust
+fn every_other_time(mut has_ran: Local<bool>) -> ShouldRun {
+    *has_ran = !*has_ran;
+    if *has_ran {
+        ShouldRun::Yes
+    } else {
+        ShouldRun::No
+    }
+}
+
+app.add_stage(SystemStage::parallel()
+   .with_system_run_criteria(every_other_time.system().label("every_other_time")))
+   .add_system(foo.system().with_run_criteria("every_other_time"))
+```
+
+Results from Run Criteria can also be "piped" into other criteria, which enables interesting composed behaviors:
+
+```rust
+fn once_in_a_blue_moon(In(input): In<ShouldRun>, moon: Res<Moon>) -> ShouldRun {
+    if moon.is_blue() {
+        input
+    } else {
+        ShouldRun::No
+    }
+}
+
+app
+    .add_system(foo.with_run_criteria(
+        "every_other_time".pipe(once_in_a_blue_moon.system())
+    )
+```
+
+### Ambiguity Detection and Resolution
+
+While the new executor is now much easier to reason about, it does introduce a new class of error: "system order ambiguities". When two systems interact with the same data, but have no explicit ordering defined, the output they produce is non-deterministic (and often not what the author intended).
+
+Consider the following app:
+
+```rust
+fn increment_counter(mut counter: ResMut<usize>) {
+    *counter += 1;
+}
+
+fn print_every_other_time(counter: Res<usize>) {
+    if *counter % 2 == 0 {
+        println!("ran");
+    }
+}
+
+app
+    .add_system(increment_counter.system())
+    .add_system(print_every_other_time.system())
+```
+
+The author clearly intended `print_every_other_time` to run every other update. However, due to the fact that these systems have no order defined, they could run in a different order each update and create a situation where nothing is printed over the course of two updates:
+
+```
+UPDATE
+- increment_counter (counter now equals 1)
+- print_every_other_time (nothing printed)
+UPDATE
+- print_every_other_time (nothing printed)
+- increment_counter (counter now equals 2)
+```
+
+The old executor would have implicitly forced `increment_counter` to run first because it conflicts with `print_every_other_time` and it was inserted first. But the new executor requires you to be explicit here (which we believe is a good thing).
+
+To help detect this class of error, we built an opt-in tool that detects these ambiguities and logs them:
+
+```
+Execution order ambiguities detected, you might want to add an explicit dependency relation between some of these systems:
+ * Parallel systems:
+ -- "&app::increment_counter" and "&app::print_every_other_time"
+    conflicts: ["usize"]
+```
```

Comment:
This should mention how the ambiguity detection is activated.

Reply:
good call